corehq/apps/data_dictionary/views.py

```diff
@@ -47,45 +65,82 @@ def data_dictionary_json(request, domain, case_type_name=None):
                 "description": prop.description,
                 "name": prop.name,
                 "data_type": prop.data_type,
+                "group": prop.group
             })
         props.append(p)
     return JsonResponse({'case_types': props})
 
 
+# atomic decorator is a performance optimization for looped saves
+# as per http://stackoverflow.com/questions/3395236/aggregating-saves-in-django#comment38715164_3397586
+@atomic
 @login_and_domain_required
 @toggles.DATA_DICTIONARY.required_decorator()
 def update_case_property(request, domain):
-    case_type = request.POST.get('caseType')
-    name = request.POST.get('name')
-    description = request.POST.get('description')
-    data_type = request.POST.get('data_type')
-    try:
-        prop = CaseProperty.objects.get(
-            name=name, case_type__name=case_type, case_type__domain=domain
-        )
-    except CaseProperty.DoesNotExist:
-        return JsonResponse({
-            'status': "failed",
-            "error": "property does not exist"
-        }, status=404)
-    if data_type:
-        prop.data_type = data_type
-    if description:
-        prop.description = description
-    try:
-        prop.full_clean()
-    except ValidationError as e:
-        return JsonResponse({"status": "failed", "error": unicode(e)}, status=400)
-    prop.save()
+    property_list = json.loads(request.POST.get('properties'))
+    for property in property_list:
+        case_type = property.get('caseType')
+        name = property.get('name')
+        description = property.get('description')
+        data_type = property.get('data_type')
+        group = property.get('group')
+        try:
+            prop = CaseProperty.objects.get(
+                name=name, case_type__name=case_type, case_type__domain=domain
+            )
+        except CaseProperty.DoesNotExist:
+            case_type_obj = CaseType.objects.get(domain=domain, name=case_type)
+            prop = CaseProperty.objects.create(case_type=case_type_obj, name=name)
```

Comment:
to prevent creating two entries for the CaseProperty on a race condition situation, we can wrap this operation in a CriticalSection